File: 2D-Arrays/20/islands.js
const directions = [
  [-1, 0], // up
  [0, 1], // right
  [1, 0], // down
  [0, -1], // left
];

var numIslands = function (grid) {
  let count = 0;
  const seen = new Array(grid.length)
    .fill(false)
    .map(() => new Array(grid[0].length).fill(false));

  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[0].length; j++) {
      if (grid[i][j] === '1') {
        count++;
        convertOldIslands(i, j, grid, seen);
      }
    }
  }
  return count;
};

var convertOldIslands = function (i, j, grid, seen) {
  if (
    i < 0 ||
    i >= grid.length ||
    j < 0 ||
    j >= grid[0].length ||
    seen[i][j] ||
    grid[i][j] == '0'
  )
    return;
  seen[i][j] = true;
  grid[i][j] = '0';
  for (let d = 0; d < directions.length; d++) {
    const currentDirection = directions[d];
    convertOldIslands(
      i + currentDirection[0],
      j + currentDirection[1],
      grid,
      seen
    );
  }
};

File: 2D-Arrays/20/islands.py
class Solution:
    def __init__(self):
        self.directions = [
            [-1, 0],
            [0, 1],
            [1, 0],
            [0, -1]
        ]

    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        seen = [[False]*len(grid[0]) for _ in range(len(grid))]
        # Sequential search
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == '1':
                    count += 1
                    self.convertOldIslands(row, col, grid, seen)
        return count
    
    # Using dfs to convert all old islands to 0
    def convertOldIslands(self, row, col, grid, seen):
        if row < 0 or row >= len(grid) or col < 0 or col >= len(grid[0]) or seen[row][col] or grid[row][col] == '0':
            return
        grid[row][col] = '0'
        seen[row][col] = True
        for direction in self.directions:
            self.convertOldIslands(row+direction[0], col+direction[1], grid, seen)

